src/core/unit.c

```diff
@@ -569,10 +569,17 @@ void unit_free(Unit *u) {
         UnitDependency d;
         Iterator i;
         char *t;
+        bool in_cleanup_queue;
 
         if (!u)
                 return;
 
+        /* Save in_cleanup_queue state, but set it it to true on the unit
+         * to prevent the unit from being added to the gc or cleanup queues
+         * during the cleanup operations. */
```

Comment:
hmm, which calls would readd it there? 

wouldn't be a nicer fix to just move the various REMOVE_LIST() calls further to the end of this function, so that the undo whatever might have been readded in the middle?

Reply:
You're right, this is a nicer solution. I reworked the patch along those lines.

So this could have a been an unpleasant bug. I added some debugging stmts, and in a container:
```
Feb 15 13:59:57 rawhide systemd[1]: Starting Power-Off...
Feb 15 13:59:57 rawhide systemd[1]: Shutting down.
Feb 15 13:59:57 rawhide systemd[1]: system-sshd\x2dkeygen.slice was added to gc queue :772
Feb 15 13:59:57 rawhide systemd[1]: system-getty.slice was added to gc queue :772
Feb 15 13:59:57 rawhide systemd[1]: dbus-broker.service was added to gc queue :772
Feb 15 13:59:57 rawhide systemd[1]: poweroff.target was added to gc queue :663
Feb 15 13:59:57 rawhide systemd-journald[17]: Journal stopped
```
Any of the lines with :772 mean that units would be left in the gc queue, which would later mean an assert would be triggered in  manager_clear_jobs_and_units(). This would happen after the journal has been shut down. I didn't do the same test without my ref loop patch, so I don't know if this would actually be triggered before it, but this seems like a potential crash-during-late-shutdown fix.